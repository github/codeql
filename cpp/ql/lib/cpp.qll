/**
 * Provides classes and predicates for working with C/C++ code.
 *
 * Where the documentation refers to the standards, it gives
 * references to the freely-available drafts.
 *
 * For C++11, these references are of the form [N3337 5.3.2/1], and the
 * corresponding draft of the standard can be downloaded from
 *     https://github.com/cplusplus/draft/raw/master/papers/n3337.pdf
 *
 * For C++14, these references are of the form [N4140 5.3.2/1], and the
 * corresponding draft of the standard can be downloaded from
 *     https://github.com/cplusplus/draft/raw/master/papers/n4140.pdf
 */

import semmle.code.cpp.File
import semmle.code.cpp.Linkage
import semmle.code.cpp.Location
import semmle.code.cpp.Compilation
import semmle.code.cpp.Element
import semmle.code.cpp.Namespace
import semmle.code.cpp.Specifier
import semmle.code.cpp.Declaration
import semmle.code.cpp.Include
import semmle.code.cpp.Macro
import semmle.code.cpp.Type
import semmle.code.cpp.TypedefType
import semmle.code.cpp.Class
import semmle.code.cpp.Struct
import semmle.code.cpp.Union
import semmle.code.cpp.Enum
import semmle.code.cpp.Member
import semmle.code.cpp.Field
import semmle.code.cpp.Function
import semmle.code.cpp.MemberFunction
import semmle.code.cpp.Parameter
import semmle.code.cpp.Variable
import semmle.code.cpp.Initializer
import semmle.code.cpp.FriendDecl
import semmle.code.cpp.exprs.Expr
import semmle.code.cpp.exprs.ArithmeticOperation
import semmle.code.cpp.exprs.BitwiseOperation
import semmle.code.cpp.exprs.LogicalOperation
import semmle.code.cpp.exprs.ComparisonOperation
import semmle.code.cpp.exprs.Assignment
import semmle.code.cpp.exprs.Cast
import semmle.code.cpp.exprs.Access
import semmle.code.cpp.exprs.Call
import semmle.code.cpp.exprs.Lambda
import semmle.code.cpp.exprs.Literal
import semmle.code.cpp.exprs.BuiltInOperations
import semmle.code.cpp.stmts.Stmt
import semmle.code.cpp.stmts.Block
import semmle.code.cpp.metrics.MetricNamespace
import semmle.code.cpp.metrics.MetricClass
import semmle.code.cpp.metrics.MetricFile
import semmle.code.cpp.metrics.MetricFunction
import semmle.code.cpp.commons.CommonType
import semmle.code.cpp.commons.Printf
import semmle.code.cpp.commons.VoidContext
import semmle.code.cpp.commons.NULL
import semmle.code.cpp.commons.PolymorphicClass
import semmle.code.cpp.commons.Alloc
import semmle.code.cpp.commons.StructLikeClass
import semmle.code.cpp.controlflow.ControlFlowGraph
import semmle.code.cpp.XML
import semmle.code.cpp.Diagnostics
import semmle.code.cpp.Comments
import semmle.code.cpp.Preprocessor
import semmle.code.cpp.Iteration
import semmle.code.cpp.NameQualifiers
import DefaultOptions
