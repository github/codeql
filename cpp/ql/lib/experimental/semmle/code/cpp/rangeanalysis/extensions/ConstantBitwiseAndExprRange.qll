private import cpp
private import experimental.semmle.code.cpp.models.interfaces.SimpleRangeAnalysisExpr
private import semmle.code.cpp.rangeanalysis.RangeAnalysisUtils

/**
 * Holds if `e` is a constant or if it is a variable with a constant value
 */
float evaluateConstantExpr(Expr e) {
  result = e.getValue().toFloat()
  or
  exists(SsaDefinition defn, StackVariable sv |
    defn.getAUse(sv) = e and
    result = defn.getDefiningValue(sv).getValue().toFloat()
  )
}

/**
 * The current implementation for `BitwiseAndExpr` only handles cases where both operands are
 * either unsigned or non-negative constants. This class not only covers these cases, but also
 * adds support for `&` expressions between a signed integer with a non-negative range and a
 * non-negative constant. It also adds support for `&=` for the same set of cases as `&`.
 */
private class ConstantBitwiseAndExprRange extends SimpleRangeAnalysisExpr {
  ConstantBitwiseAndExprRange() {
    exists(Expr l, Expr r |
      l = this.(BitwiseAndExpr).getLeftOperand() and
      r = this.(BitwiseAndExpr).getRightOperand()
      or
      l = this.(AssignAndExpr).getLValue() and
      r = this.(AssignAndExpr).getRValue()
    |
      // No operands can be negative constants
      not (evaluateConstantExpr(l) < 0 or evaluateConstantExpr(r) < 0) and
      // At least one operand must be a non-negative constant
      (evaluateConstantExpr(l) >= 0 or evaluateConstantExpr(r) >= 0)
    )
  }

  Expr getLeftOperand() {
    result = this.(BitwiseAndExpr).getLeftOperand() or
    result = this.(AssignAndExpr).getLValue()
  }

  Expr getRightOperand() {
    result = this.(BitwiseAndExpr).getRightOperand() or
    result = this.(AssignAndExpr).getRValue()
  }

  override QlBuiltins::BigInt getLowerBounds() {
    // If an operand can have negative values, the lower bound is unconstrained.
    // Otherwise, the lower bound is zero.
    exists(QlBuiltins::BigInt lLower, QlBuiltins::BigInt rLower |
      lLower = getFullyConvertedLowerBounds(this.getLeftOperand()) and
      rLower = getFullyConvertedLowerBounds(this.getRightOperand()) and
      (
        (lLower < 0.toBigInt() or rLower < 0.toBigInt()) and
        result = exprMinVal(this)
        or
        // This technically results in two lowerBounds when an operand range is negative, but
        // that's fine since `exprMinVal(x) <= 0`. We can't use an if statement here without
        // non-monotonic recursion issues
        result = 0.toBigInt()
      )
    )
  }

  override QlBuiltins::BigInt getUpperBounds() {
    // If an operand can have negative values, the upper bound is unconstrained.
    // Otherwise, the upper bound is the minimum of the upper bounds of the operands
    exists(
      QlBuiltins::BigInt lLower, QlBuiltins::BigInt lUpper, QlBuiltins::BigInt rLower,
      QlBuiltins::BigInt rUpper
    |
      lLower = getFullyConvertedLowerBounds(this.getLeftOperand()) and
      lUpper = getFullyConvertedUpperBounds(this.getLeftOperand()) and
      rLower = getFullyConvertedLowerBounds(this.getRightOperand()) and
      rUpper = getFullyConvertedUpperBounds(this.getRightOperand()) and
      (
        (lLower < 0.toBigInt() or rLower < 0.toBigInt()) and
        result = exprMaxVal(this)
        or
        // This technically results in two upperBounds when an operand range is negative, but
        // that's fine since `exprMaxVal(b) >= result`. We can't use an if statement here without
        // non-monotonic recursion issues
        result = rUpper.minimum(lUpper)
      )
    )
  }

  override predicate dependsOnChild(Expr child) {
    child = this.getLeftOperand() or child = this.getRightOperand()
  }
}
