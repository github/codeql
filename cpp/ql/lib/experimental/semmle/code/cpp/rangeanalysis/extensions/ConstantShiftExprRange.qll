private import cpp
private import experimental.semmle.code.cpp.models.interfaces.SimpleRangeAnalysisExpr
private import semmle.code.cpp.rangeanalysis.RangeAnalysisUtils

float evaluateConstantExpr(Expr e) {
  result = e.getValue().toFloat()
  or
  // This handles when a constant value is put into a variable
  // and the variable is used later
  exists(SsaDefinition defn, StackVariable sv |
    defn.getAUse(sv) = e and
    result = defn.getDefiningValue(sv).getValue().toFloat()
  )
}

// If the constant right operand is negative or is greater than or equal to the number of
// bits in the left operands type, then the result is undefined (except on the IA-32
// architecture where the shift value is masked with 0b00011111, but we can't
// assume the architecture).
bindingset[val]
private predicate isValidShiftExprShift(float val, Expr l) {
  val >= 0 and
  // We use getFullyConverted because the spec says to use the *promoted* left operand
  val < (l.getFullyConverted().getUnderlyingType().getSize() * 8)
}

bindingset[val, shift, max_val]
private predicate canLShiftOverflow(int val, int shift, int max_val) {
  // val << shift = val * 2^shift > max_val => val > max_val/2^shift = max_val >> b
  val > max_val.bitShiftRight(shift)
}

/**
 * A range analysis expression consisting of the `>>` or `>>=` operator when at least
 * one operand is a constant (and if the right operand is a constant, it must be "valid"
 * (see `isValidShiftExprShift`)). When handling any undefined behavior, it leaves the
 * values unconstrained. From the C++ standard: "The behavior is undefined if the right
 * operand is negative, or greater than or equal to the length in bits of the promoted
 * left operand. The value of E1 >> E2 is E1 right-shifted E2 bit positions. If E1 has an
 * unsigned type or if E1 has a signed type and a non-negative value, the value of the
 * result is the integral part of the quotient of E1/2^E2. If E1 has a signed type and a
 * negative value, the resulting value is implementation-defined."
 */
class ConstantRShiftExprRange extends SimpleRangeAnalysisExpr {
  /**
   * Holds for `a >> b` or `a >>= b` in one of the following two cases:
   * 1. `a` is a constant and `b` is not
   * 2. `b` is constant
   *
   * We don't handle the case where `a` and `b` are both non-constant values.
   */
  ConstantRShiftExprRange() {
    getUnspecifiedType() instanceof IntegralType and
    exists(Expr l, Expr r |
      l = this.(RShiftExpr).getLeftOperand() and
      r = this.(RShiftExpr).getRightOperand()
      or
      l = this.(AssignRShiftExpr).getLValue() and
      r = this.(AssignRShiftExpr).getRValue()
    |
      l.getUnspecifiedType() instanceof IntegralType and
      r.getUnspecifiedType() instanceof IntegralType and
      (
        // If the left operand is a constant, verify that the right operand is not a constant
        exists(evaluateConstantExpr(l)) and not exists(evaluateConstantExpr(r))
        or
        // If the right operand is a constant, check if it is a valid shift expression
        exists(float constROp |
          constROp = evaluateConstantExpr(r) and isValidShiftExprShift(constROp, l)
        )
      )
    )
  }

  Expr getLeftOperand() {
    result = this.(RShiftExpr).getLeftOperand() or
    result = this.(AssignRShiftExpr).getLValue()
  }

  Expr getRightOperand() {
    result = this.(RShiftExpr).getRightOperand() or
    result = this.(AssignRShiftExpr).getRValue()
  }

  override float getLowerBounds() {
    exists(int lLower, int lUpper, int rLower, int rUpper |
      lLower = getFullyConvertedLowerBounds(getLeftOperand()) and
      lUpper = getFullyConvertedUpperBounds(getLeftOperand()) and
      rLower = getFullyConvertedLowerBounds(getRightOperand()) and
      rUpper = getFullyConvertedUpperBounds(getRightOperand()) and
      lLower <= lUpper and
      rLower <= rUpper
    |
      if
        lLower < 0
        or
        not (
          isValidShiftExprShift(rLower, getLeftOperand()) and
          isValidShiftExprShift(rUpper, getLeftOperand())
        )
      then
        // We don't want to deal with shifting negative numbers at the moment,
        // and a negative shift is implementation defined, so we set the result
        // to the minimum value
        result = exprMinVal(this)
      else
        // We can get the smallest value by shifting the smallest bound by the largest bound
        result = lLower.bitShiftRight(rUpper)
    )
  }

  override float getUpperBounds() {
    exists(int lLower, int lUpper, int rLower, int rUpper |
      lLower = getFullyConvertedLowerBounds(getLeftOperand()) and
      lUpper = getFullyConvertedUpperBounds(getLeftOperand()) and
      rLower = getFullyConvertedLowerBounds(getRightOperand()) and
      rUpper = getFullyConvertedUpperBounds(getRightOperand()) and
      lLower <= lUpper and
      rLower <= rUpper
    |
      if
        lLower < 0
        or
        not (
          isValidShiftExprShift(rLower, getLeftOperand()) and
          isValidShiftExprShift(rUpper, getLeftOperand())
        )
      then
        // We don't want to deal with shifting negative numbers at the moment,
        // and a negative shift is implementation defined, so we set the result
        // to the maximum value
        result = exprMaxVal(this)
      else
        // We can get the largest value by shifting the largest bound by the smallest bound
        result = lUpper.bitShiftRight(rLower)
    )
  }

  override predicate dependsOnChild(Expr child) {
    child = getLeftOperand() or child = getRightOperand()
  }
}

/**
 * A range analysis expression consisting of the `<<` or `<<=` operator when at least
 * one operand is a constant (and if the right operand is a constant, it must be "valid"
 * (see `isValidShiftExprShift`)). When handling any undefined behavior, it leaves the
 * values unconstrained. From the C++ standard: "The behavior is undefined if the right
 * operand is negative, or greater than or equal to the length in bits of the promoted left operand.
 * The value of E1 << E2 is E1 left-shifted E2 bit positions; vacated bits are zero-filled. If E1
 * has an unsigned type, the value of the result is E1 x 2 E2, reduced modulo one more than the
 * maximum value representable in the result type. Otherwise, if E1 has a signed type and
 * non-negative value, and E1 x 2 E2 is representable in the corresponding unsigned type of the
 * result type, then that value, converted to the result type, is the resulting value; otherwise,
 * the behavior is undefined."
 */
class ConstantLShiftExprRange extends SimpleRangeAnalysisExpr {
  /**
   * Holds for `a << b` or `a <<= b` in one of the following two cases:
   * 1. `a` is a constant and `b` is not
   * 2. `b` is constant
   *
   * We don't handle the case where `a` and `b` are both non-constant values.
   */
  ConstantLShiftExprRange() {
    getUnspecifiedType() instanceof IntegralType and
    exists(Expr l, Expr r |
      l = this.(LShiftExpr).getLeftOperand() and
      r = this.(LShiftExpr).getRightOperand()
      or
      l = this.(AssignLShiftExpr).getLValue() and
      r = this.(AssignLShiftExpr).getRValue()
    |
      l.getUnspecifiedType() instanceof IntegralType and
      r.getUnspecifiedType() instanceof IntegralType and
      (
        // If the left operand is a constant, verify that the right operand is not a constant
        exists(evaluateConstantExpr(l)) and not exists(evaluateConstantExpr(r))
        or
        // If the right operand is a constant, check if it is a valid shift expression
        exists(float constROp |
          constROp = evaluateConstantExpr(r) and isValidShiftExprShift(constROp, l)
        )
      )
    )
  }

  Expr getLeftOperand() {
    result = this.(LShiftExpr).getLeftOperand() or
    result = this.(AssignLShiftExpr).getLValue()
  }

  Expr getRightOperand() {
    result = this.(LShiftExpr).getRightOperand() or
    result = this.(AssignLShiftExpr).getRValue()
  }

  override float getLowerBounds() {
    exists(int lLower, int lUpper, int rLower, int rUpper |
      lLower = getFullyConvertedLowerBounds(getLeftOperand()) and
      lUpper = getFullyConvertedUpperBounds(getLeftOperand()) and
      rLower = getFullyConvertedLowerBounds(getRightOperand()) and
      rUpper = getFullyConvertedUpperBounds(getRightOperand()) and
      lLower <= lUpper and
      rLower <= rUpper
    |
      if
        lLower < 0
        or
        not (
          isValidShiftExprShift(rLower, getLeftOperand()) and
          isValidShiftExprShift(rUpper, getLeftOperand())
        )
      then
        // We don't want to deal with shifting negative numbers at the moment,
        // and a negative shift is undefined, so we set to the minimum value
        result = exprMinVal(this)
      else
        // If we have `0b01010000 << [0, 2]`, the max value for 8 bits is 0b10100000
        // (a shift of 1) but doing a shift by the upper bound would give 0b01000000.
        // So if the left shift operation causes an overflow, we just assume the max value
        // If necessary, we may be able to improve this bound in the future
        if canLShiftOverflow(lUpper, rUpper, exprMaxVal(this))
        then result = exprMinVal(this)
        else result = lLower.bitShiftLeft(rLower)
    )
  }

  override float getUpperBounds() {
    exists(int lLower, int lUpper, int rLower, int rUpper |
      lLower = getFullyConvertedLowerBounds(getLeftOperand()) and
      lUpper = getFullyConvertedUpperBounds(getLeftOperand()) and
      rLower = getFullyConvertedLowerBounds(getRightOperand()) and
      rUpper = getFullyConvertedUpperBounds(getRightOperand()) and
      lLower <= lUpper and
      rLower <= rUpper
    |
      if
        lLower < 0
        or
        not (
          isValidShiftExprShift(rLower, getLeftOperand()) and
          isValidShiftExprShift(rUpper, getLeftOperand())
        )
      then
        // We don't want to deal with shifting negative numbers at the moment,
        // and a negative shift is undefined, so we set it to the maximum value
        result = exprMaxVal(this)
      else
        // If we have `0b01010000 << [0, 2]`, the max value for 8 bits is 0b10100000
        // (a shift of 1) but doing a shift by the upper bound would give 0b01000000.
        // So if the left shift operation causes an overflow, we just assume the max value
        // If necessary, we may be able to improve this bound in the future
        if canLShiftOverflow(lUpper, rUpper, exprMaxVal(this))
        then result = exprMaxVal(this)
        else result = lUpper.bitShiftLeft(rUpper)
    )
  }

  override predicate dependsOnChild(Expr child) {
    child = getLeftOperand() or child = getRightOperand()
  }
}
