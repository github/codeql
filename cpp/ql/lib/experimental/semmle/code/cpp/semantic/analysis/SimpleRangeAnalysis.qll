/**
 * Wrapper for the semantic range analysis library that mimics the
 * interface of the simple range analysis library.
 */

private import cpp
private import semmle.code.cpp.ir.IR
private import experimental.semmle.code.cpp.semantic.SemanticBound
private import experimental.semmle.code.cpp.semantic.SemanticExprSpecific
private import RangeAnalysis

/**
 * Gets the lower bound of the expression.
 *
 * Note: expressions in C/C++ are often implicitly or explicitly cast to a
 * different result type. Such casts can cause the value of the expression
 * to overflow or to be truncated. This predicate computes the lower bound
 * of the expression without including the effect of the casts. To compute
 * the lower bound of the expression after all the casts have been applied,
 * call `lowerBound` like this:
 *
 *    `lowerBound(expr.getFullyConverted())`
 */
float lowerBound(Expr expr) {
  exists(Instruction i, SemBound b | i.getAst() = expr and b instanceof SemZeroBound |
    semBounded(getSemanticExpr(i), b, result, false, _)
  )
}

/**
 * Gets the upper bound of the expression.
 *
 * Note: expressions in C/C++ are often implicitly or explicitly cast to a
 * different result type. Such casts can cause the value of the expression
 * to overflow or to be truncated. This predicate computes the upper bound
 * of the expression without including the effect of the casts. To compute
 * the upper bound of the expression after all the casts have been applied,
 * call `upperBound` like this:
 *
 *    `upperBound(expr.getFullyConverted())`
 */
float upperBound(Expr expr) {
  exists(Instruction i, SemBound b | i.getAst() = expr and b instanceof SemZeroBound |
    semBounded(getSemanticExpr(i), b, result, true, _)
  )
}

/**
 * Holds if the upper bound of `expr` may have been widened. This means the
 * upper bound is in practice likely to be overly wide.
 */
predicate upperBoundMayBeWidened(Expr e) { none() }

/**
 * Holds if `expr` has a provably empty range. For example:
 *
 *   10 < expr and expr < 5
 *
 * The range of an expression can only be empty if it can never be
 * executed. For example:
 *
 * ```cpp
 * if (10 < x) {
 *   if (x < 5) {
 *     // Unreachable code
 *     return x; // x has an empty range: 10 < x && x < 5
 *   }
 * }
 * ```
 */
predicate exprWithEmptyRange(Expr expr) { lowerBound(expr) > upperBound(expr) }

/** Holds if the definition might overflow negatively. */
predicate defMightOverflowNegatively(RangeSsaDefinition def, StackVariable v) { none() }

/** Holds if the definition might overflow positively. */
predicate defMightOverflowPositively(RangeSsaDefinition def, StackVariable v) { none() }

/**
 * Holds if the definition might overflow (either positively or
 * negatively).
 */
predicate defMightOverflow(RangeSsaDefinition def, StackVariable v) {
  defMightOverflowNegatively(def, v) or
  defMightOverflowPositively(def, v)
}

/**
 * Holds if the expression might overflow negatively. This predicate
 * does not consider the possibility that the expression might overflow
 * due to a conversion.
 */
predicate exprMightOverflowNegatively(Expr expr) { none() }

/**
 * Holds if the expression might overflow negatively. Conversions
 * are also taken into account. For example the expression
 * `(int16)(x+y)` might overflow due to the `(int16)` cast, rather than
 * due to the addition.
 */
predicate convertedExprMightOverflowNegatively(Expr expr) {
  exprMightOverflowNegatively(expr) or
  convertedExprMightOverflowNegatively(expr.getConversion())
}

/**
 * Holds if the expression might overflow positively. This predicate
 * does not consider the possibility that the expression might overflow
 * due to a conversion.
 */
predicate exprMightOverflowPositively(Expr expr) { none() }

/**
 * Holds if the expression might overflow positively. Conversions
 * are also taken into account. For example the expression
 * `(int16)(x+y)` might overflow due to the `(int16)` cast, rather than
 * due to the addition.
 */
predicate convertedExprMightOverflowPositively(Expr expr) {
  exprMightOverflowPositively(expr) or
  convertedExprMightOverflowPositively(expr.getConversion())
}

/**
 * Holds if the expression might overflow (either positively or
 * negatively). The possibility that the expression might overflow
 * due to an implicit or explicit cast is also considered.
 */
predicate convertedExprMightOverflow(Expr expr) {
  convertedExprMightOverflowNegatively(expr) or
  convertedExprMightOverflowPositively(expr)
}
