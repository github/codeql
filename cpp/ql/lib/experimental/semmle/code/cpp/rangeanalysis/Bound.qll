import semmle.code.cpp.rangeanalysis.new.internal.semantic.analysis.Bound
