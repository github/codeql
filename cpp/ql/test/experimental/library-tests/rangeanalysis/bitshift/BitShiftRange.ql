import cpp
import semmle.code.cpp.rangeanalysis.SimpleRangeAnalysis
import experimental.semmle.code.cpp.rangeanalysis.extensions.ConstantShiftExprRange

Expr getLOp(Operation o) {
  result = o.(BinaryOperation).getLeftOperand() or
  result = o.(Assignment).getLValue()
}

Expr getROp(Operation o) {
  result = o.(BinaryOperation).getRightOperand() or
  result = o.(Assignment).getRValue()
}

from Operation o
where
  (
    o instanceof BinaryBitwiseOperation
    or
    o instanceof AssignBitwiseOperation
  )
select o, lowerBound(o), upperBound(o), getLOp(o).getUnderlyingType(),
  getROp(o).getUnderlyingType(), getLOp(o).getFullyConverted().getUnderlyingType(),
  getROp(o).getFullyConverted().getUnderlyingType()
