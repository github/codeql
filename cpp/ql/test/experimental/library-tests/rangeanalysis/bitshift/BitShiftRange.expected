| bitshift.cpp:23:3:23:9 | ... <<= ... | 0.0 | 255.0 | file://:0:0:0:0 | unsigned char | file://:0:0:0:0 | int | file://:0:0:0:0 | unsigned char | file://:0:0:0:0 | int |
| bitshift.cpp:25:5:25:11 | ... <<= ... | 0.0 | 240.0 | file://:0:0:0:0 | unsigned char | file://:0:0:0:0 | int | file://:0:0:0:0 | unsigned char | file://:0:0:0:0 | int |
| bitshift.cpp:29:3:29:8 | ... << ... | 0.0 | 1020.0 | file://:0:0:0:0 | unsigned char | file://:0:0:0:0 | int | file://:0:0:0:0 | int | file://:0:0:0:0 | int |
| bitshift.cpp:32:3:32:9 | ... << ... | -2.147483648E9 | 2.147483647E9 | file://:0:0:0:0 | unsigned char | file://:0:0:0:0 | int | file://:0:0:0:0 | int | file://:0:0:0:0 | int |
| bitshift.cpp:35:3:35:9 | ... << ... | -2.147483648E9 | 2.147483647E9 | file://:0:0:0:0 | unsigned char | file://:0:0:0:0 | int | file://:0:0:0:0 | int | file://:0:0:0:0 | int |
| bitshift.cpp:38:3:38:22 | ... << ... | 0.0 | 32640.0 | file://:0:0:0:0 | unsigned char | file://:0:0:0:0 | unsigned char | file://:0:0:0:0 | int | file://:0:0:0:0 | int |
| bitshift.cpp:39:3:39:22 | ... << ... | 0.0 | 32640.0 | file://:0:0:0:0 | unsigned char | file://:0:0:0:0 | unsigned char | file://:0:0:0:0 | int | file://:0:0:0:0 | int |
| bitshift.cpp:40:3:40:22 | ... << ... | 0.0 | 32640.0 | file://:0:0:0:0 | unsigned char | file://:0:0:0:0 | unsigned char | file://:0:0:0:0 | int | file://:0:0:0:0 | int |
| bitshift.cpp:43:3:43:19 | ... << ... | -2.147483648E9 | 2.147483647E9 | file://:0:0:0:0 | unsigned char | file://:0:0:0:0 | signed char | file://:0:0:0:0 | int | file://:0:0:0:0 | int |
| bitshift.cpp:46:3:46:22 | ... << ... | 128.0 | 128.0 | file://:0:0:0:0 | int | file://:0:0:0:0 | unsigned char | file://:0:0:0:0 | int | file://:0:0:0:0 | int |
| bitshift.cpp:49:3:49:8 | ... << ... | -2.147483648E9 | 2.147483647E9 | file://:0:0:0:0 | int | file://:0:0:0:0 | unsigned char | file://:0:0:0:0 | int | file://:0:0:0:0 | int |
| bitshift.cpp:52:5:52:10 | ... << ... | 1.0 | 128.0 | file://:0:0:0:0 | int | file://:0:0:0:0 | unsigned char | file://:0:0:0:0 | int | file://:0:0:0:0 | int |
| bitshift.cpp:57:3:57:8 | ... << ... | -2.147483648E9 | 2.147483647E9 | file://:0:0:0:0 | signed char | file://:0:0:0:0 | int | file://:0:0:0:0 | int | file://:0:0:0:0 | int |
| bitshift.cpp:58:3:58:9 | ... << ... | -2.147483648E9 | 2.147483647E9 | file://:0:0:0:0 | signed char | file://:0:0:0:0 | int | file://:0:0:0:0 | int | file://:0:0:0:0 | int |
| bitshift.cpp:59:3:59:9 | ... << ... | -2.147483648E9 | 2.147483647E9 | file://:0:0:0:0 | signed char | file://:0:0:0:0 | int | file://:0:0:0:0 | int | file://:0:0:0:0 | int |
| bitshift.cpp:60:3:60:22 | ... << ... | -2.147483648E9 | 2.147483647E9 | file://:0:0:0:0 | signed char | file://:0:0:0:0 | unsigned char | file://:0:0:0:0 | int | file://:0:0:0:0 | int |
| bitshift.cpp:61:3:61:19 | ... << ... | -2.147483648E9 | 2.147483647E9 | file://:0:0:0:0 | signed char | file://:0:0:0:0 | signed char | file://:0:0:0:0 | int | file://:0:0:0:0 | int |
| bitshift.cpp:64:3:64:19 | ... << ... | -2.147483648E9 | 2.147483647E9 | file://:0:0:0:0 | int | file://:0:0:0:0 | signed char | file://:0:0:0:0 | int | file://:0:0:0:0 | int |
| bitshift.cpp:67:3:67:8 | ... << ... | -2.147483648E9 | 2.147483647E9 | file://:0:0:0:0 | int | file://:0:0:0:0 | signed char | file://:0:0:0:0 | int | file://:0:0:0:0 | int |
| bitshift.cpp:70:5:70:10 | ... << ... | 1.0 | 128.0 | file://:0:0:0:0 | int | file://:0:0:0:0 | signed char | file://:0:0:0:0 | int | file://:0:0:0:0 | int |
| bitshift.cpp:75:5:75:10 | ... << ... | -2.147483648E9 | 2.147483647E9 | file://:0:0:0:0 | unsigned char | file://:0:0:0:0 | signed char | file://:0:0:0:0 | int | file://:0:0:0:0 | int |
| bitshift.cpp:76:5:76:10 | ... << ... | -2.147483648E9 | 2.147483647E9 | file://:0:0:0:0 | signed char | file://:0:0:0:0 | unsigned char | file://:0:0:0:0 | int | file://:0:0:0:0 | int |
| bitshift.cpp:90:3:90:9 | ... >>= ... | 0.0 | 63.0 | file://:0:0:0:0 | unsigned char | file://:0:0:0:0 | int | file://:0:0:0:0 | unsigned char | file://:0:0:0:0 | int |
| bitshift.cpp:92:5:92:11 | ... >>= ... | 0.0 | 15.0 | file://:0:0:0:0 | unsigned char | file://:0:0:0:0 | int | file://:0:0:0:0 | unsigned char | file://:0:0:0:0 | int |
| bitshift.cpp:96:3:96:8 | ... >> ... | 0.0 | 63.0 | file://:0:0:0:0 | unsigned char | file://:0:0:0:0 | int | file://:0:0:0:0 | int | file://:0:0:0:0 | int |
| bitshift.cpp:99:3:99:9 | ... >> ... | 0.0 | 0.0 | file://:0:0:0:0 | unsigned char | file://:0:0:0:0 | int | file://:0:0:0:0 | int | file://:0:0:0:0 | int |
| bitshift.cpp:103:3:103:9 | ... >> ... | 0.0 | 0.0 | file://:0:0:0:0 | unsigned char | file://:0:0:0:0 | int | file://:0:0:0:0 | int | file://:0:0:0:0 | int |
| bitshift.cpp:106:3:106:22 | ... >> ... | 0.0 | 63.0 | file://:0:0:0:0 | unsigned char | file://:0:0:0:0 | unsigned char | file://:0:0:0:0 | int | file://:0:0:0:0 | int |
| bitshift.cpp:107:3:107:22 | ... >> ... | 0.0 | 63.0 | file://:0:0:0:0 | unsigned char | file://:0:0:0:0 | unsigned char | file://:0:0:0:0 | int | file://:0:0:0:0 | int |
| bitshift.cpp:108:3:108:22 | ... >> ... | 0.0 | 63.0 | file://:0:0:0:0 | unsigned char | file://:0:0:0:0 | unsigned char | file://:0:0:0:0 | int | file://:0:0:0:0 | int |
| bitshift.cpp:111:3:111:19 | ... >> ... | -2.147483648E9 | 2.147483647E9 | file://:0:0:0:0 | unsigned char | file://:0:0:0:0 | signed char | file://:0:0:0:0 | int | file://:0:0:0:0 | int |
| bitshift.cpp:114:3:114:24 | ... >> ... | 32.0 | 32.0 | file://:0:0:0:0 | int | file://:0:0:0:0 | unsigned char | file://:0:0:0:0 | int | file://:0:0:0:0 | int |
| bitshift.cpp:117:3:117:10 | ... >> ... | -2.147483648E9 | 2.147483647E9 | file://:0:0:0:0 | int | file://:0:0:0:0 | unsigned char | file://:0:0:0:0 | int | file://:0:0:0:0 | int |
| bitshift.cpp:120:5:120:12 | ... >> ... | 32.0 | 128.0 | file://:0:0:0:0 | int | file://:0:0:0:0 | unsigned char | file://:0:0:0:0 | int | file://:0:0:0:0 | int |
| bitshift.cpp:126:3:126:8 | ... >> ... | -2.147483648E9 | 2.147483647E9 | file://:0:0:0:0 | signed char | file://:0:0:0:0 | int | file://:0:0:0:0 | int | file://:0:0:0:0 | int |
| bitshift.cpp:127:3:127:9 | ... >> ... | -2.147483648E9 | 2.147483647E9 | file://:0:0:0:0 | signed char | file://:0:0:0:0 | int | file://:0:0:0:0 | int | file://:0:0:0:0 | int |
| bitshift.cpp:128:3:128:9 | ... >> ... | -1.0 | 0.0 | file://:0:0:0:0 | signed char | file://:0:0:0:0 | int | file://:0:0:0:0 | int | file://:0:0:0:0 | int |
| bitshift.cpp:129:3:129:22 | ... >> ... | -2.147483648E9 | 2.147483647E9 | file://:0:0:0:0 | signed char | file://:0:0:0:0 | unsigned char | file://:0:0:0:0 | int | file://:0:0:0:0 | int |
| bitshift.cpp:130:3:130:19 | ... >> ... | -2.147483648E9 | 2.147483647E9 | file://:0:0:0:0 | signed char | file://:0:0:0:0 | signed char | file://:0:0:0:0 | int | file://:0:0:0:0 | int |
| bitshift.cpp:133:3:133:21 | ... >> ... | -2.147483648E9 | 2.147483647E9 | file://:0:0:0:0 | int | file://:0:0:0:0 | signed char | file://:0:0:0:0 | int | file://:0:0:0:0 | int |
| bitshift.cpp:136:3:136:10 | ... >> ... | -2.147483648E9 | 2.147483647E9 | file://:0:0:0:0 | int | file://:0:0:0:0 | signed char | file://:0:0:0:0 | int | file://:0:0:0:0 | int |
| bitshift.cpp:139:5:139:12 | ... >> ... | 32.0 | 128.0 | file://:0:0:0:0 | int | file://:0:0:0:0 | signed char | file://:0:0:0:0 | int | file://:0:0:0:0 | int |
| bitshift.cpp:144:5:144:10 | ... >> ... | -2.147483648E9 | 2.147483647E9 | file://:0:0:0:0 | unsigned char | file://:0:0:0:0 | signed char | file://:0:0:0:0 | int | file://:0:0:0:0 | int |
| bitshift.cpp:145:5:145:10 | ... >> ... | -2.147483648E9 | 2.147483647E9 | file://:0:0:0:0 | signed char | file://:0:0:0:0 | unsigned char | file://:0:0:0:0 | int | file://:0:0:0:0 | int |
